### PR TITLE
Allow specifying provided polyfills (#2)

### PR DESCRIPTION
**Pull request by sobstel**
Allow to whitelist particular non-syntax features as polyfills, so they aren't reported as problems.

```json
{
  "rules": {
    "ecmascript-compat/compat": [
      "error",
      {
        "polyfills": [
          "Array.prototype.flat"
        ]
      }
    ]
  }
}
```

### CODE REVIEW

File: packages/eslint-plugin-ecmascript-compat/lib/delegation.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/delegation.js]: Note from `package.json:engines` (the non-root ones) that Node 14 is required, so optional chaining and nullish coalescing operators are available

File: packages/eslint-plugin-ecmascript-compat/lib/rule.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/rule.js]: The spellings are "polyfills" and "polyfill" for a single one.

It's from [Polyfilla](https://en.wikipedia.org/wiki/Spackling_paste#Polyfilla) by the way :); we are filling holes in the browser.

**Comment by robatwilliams** [comment]
Hi, thanks for taking a look at this.

In principle I don't agree with doing string inclusion checks to achieve this.

How about something like this?

1. On every rule definition, add an optional property `polyfill` property with the text that must go in the rule options.
2. In `rule.js:create()`, filter out the rules.

Once past draft state, also:
1. Add unit tests for every polyfillable rule under `valid:` in the `es2xxx.spec.js` files (see [here](https://eslint.org/docs/latest/developer-guide/nodejs-api#ruletester) for how to specify options). 
2. Schema should validate that only known polyfills are specified. The valid strings _could_ be pulled off the rule definitions programmatically, but it's perhaps better just to hard code all the strings here so it can serve as documentation.
3. Document in readme (I'm happy to do that)

**Comment by sobstel** [comment]
Thanks a lot for quick and very detailed feedback. I found checking inclusions very tempting (mostly due to easy implementation) but no doubt, your  suggestion is much more solid and bullet-proof. I actually had similar idea with expanding rules, but ruled it out at the beginning. I thought it might be a little too complex, but wrongly so. I'll let you know when I have it improved.

**Comment by sobstel** [comment]
> 1. On every rule definition, add an optional property `polyfill` property with the text that must go in the rule options.
> 2. In `rule.js:create()`, filter out the rules.

Both done. @robatwilliams Could you please have a look if we're on the same page regarding implementation details?

Usage:

```json
{
      " ecmascript-compat/compat": [
         " error",
         {
           " polyfills":[
                 "Array.prototype.flat",
                 "Array.prototype.flatMap",
                 "Array.prototype.includes",
                 "Promise.prototype.finally",
                 "Object.entries",
                 "Object.fromEntries",
                 "Object.getOwnPropertyDescriptors",
                 "Object.values",
                 "String.prototype.padStart",
                 " String.prototype.replaceAll",
                 " String.prototype.trimEnd",
                 " String.prototype.trimStart",
              ],
```
 
> Once past draft state, also: (...)

I'll add unit tests and update schema once we agree on above.

File: packages/eslint-plugin-ecmascript-compat/lib/features/es2021.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/features/es2021.js:16]: `Promise.any` below should also be polyfillable.

I checked against this [list](https://github.com/robatwilliams/es-compat/blob/master/packages/eslint-plugin-ecmascript-compat/lib/features/es-versions.md) and you've got all the others.

File: packages/eslint-plugin-ecmascript-compat/lib/rule.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/rule.js]: I think we should rename this function and the var, because it's not necessarily forbidden now.

I suggest `unsupportedFeatures`

File: packages/eslint-plugin-ecmascript-compat/lib/rule.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/rule.js:48]: `context` is always given, so we don't need to use optional chaining when accessing `options` on it

File: packages/eslint-plugin-ecmascript-compat/lib/polyfilling.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/polyfilling.js]: In another comment I proposed an `isPolyfilled()` function instead.

Like this one, it should be small enough that it can live in `rule.js`

**Review by robatwilliams — CHANGES_REQUESTED** [review]

**Comment by robatwilliams** [comment]
Looks good, thanks.

I looked to see if I could allow the workflow to run without my approval, without changing the setting for all potential contributors - but there isn't anything. You can run it locally via `npm run ci`, or more specific tasks you'll find in `package.json`.

File: packages/eslint-plugin-ecmascript-compat/lib/polyfilling.js
Comment (sobstel) [reply packages/eslint-plugin-ecmascript-compat/lib/polyfilling.js]: >  it should be small enough that it can live in rule.js

I'm future-proofing here as I thought about moving polyfill strings to constants (enum) that can be re-used when defining schema. I wanted to put them into `polyfilling.js` file.

File: packages/eslint-plugin-ecmascript-compat/lib/rule.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/rule.js]: Apologies, I wrote a comment here before, but it looks like I didn't post it :(

I don't think the concept of "valid feature" is necessary, and we could do without some intermediate variables.

How about something like this to express the intent?

```
const forbiddenFeatures = unsupportedFeatures
  .filter(feature => !polyfilling.isPolyfilled(feature, polyfills))
const visitors = forbiddenFeatures .map((feature) => createDelegatee(feature.ruleConfig, context));
```

File: packages/eslint-plugin-ecmascript-compat/lib/rule.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/rule.js]: Would be good to normalise this (`?? []`) here at the point it comes into our code, instead of needing `?.` in `isPolyfilled`

**Review by robatwilliams — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin-ecmascript-compat/lib/polyfilling.js
Comment (robatwilliams) [reply packages/eslint-plugin-ecmascript-compat/lib/polyfilling.js]: I see. There is this from before though:

> Schema should validate that only known polyfills are specified. The valid strings could be pulled off the rule definitions programmatically, but it's perhaps better just to hard code all the strings here so it can serve as documentation.

By documentation, I mean for users using the plugin, rather than for developers of this plugin. I don't think they should have to follow constants through the codebase to find out the accepted polyfills - they should just be able to look at the schema. The draft schema you posted earlier looks good for that.

Comment (sobstel) [reply packages/eslint-plugin-ecmascript-compat/lib/polyfilling.js]: Ok. Makes sense to me. Let's remove polyfiling.js file. It's simple call and truth to be told it doesn't really deserve to be a separate func call on its own.

File: packages/eslint-plugin-ecmascript-compat/lib/rule.js
Comment (sobstel) [reply packages/eslint-plugin-ecmascript-compat/lib/rule.js]: Agreed. Should be much more readable. I've also got rid of polyfilling file and also isPolyfilled func (I think there's no much of readability gain in having a wrapper around simple array includes call after all).

**Comment by sobstel** [comment]
> You can run it locally via `npm run ci` (...)

It should pass now (at least it passes locally after most recent fixes).

File: packages/eslint-plugin-ecmascript-compat/lib/rule.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/rule.js]: Let's do without this variable I had, and just do `feature.ruleConfig` when calling `createDelegatee()`

**Comment by sobstel** [comment]
Thanks for updating docs @robatwilliams. I'll update the schema and unit tests.

**Comment by sobstel** [comment]
@robatwilliams I have added unit tests for all polyfills and updated json schema. It's ready for final review now.

```
> ci
> npm-run-all check test


> check
> npm-run-all --parallel check:*


> check:lint
> eslint --no-eslintrc --config=.eslintrc.json **/*.js


> check:prettier
> prettier --check **/*.{js,json,md}

Checking formatting...
All matched files use Prettier code style!

> test
> lerna run test

lerna notice cli v4.0.0
lerna info Executing command in 1 package: "npm run test"
lerna info run Ran npm script 'test' in 'eslint-plugin-ecmascript-compat' in 7.5s:

> eslint-plugin-ecmascript-compat@2.1.0 test
> jest --env=node

lerna success run Ran npm script 'test' in 1 package in 7.5s:
lerna success - eslint-plugin-ecmascript-compat
```

File: packages/eslint-plugin-ecmascript-compat/lib/features/es2016.spec.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/features/es2016.spec.js]: For the purpose of what we're testing, I don't think it's necessary to test both the direct prototype usage here in addition to the prototype traversal usage above. One of the two is enough.

(applies to many files - I won't duplicate this comment

File: packages/eslint-plugin-ecmascript-compat/lib/features/es2020.spec.js
Comment (robatwilliams) [inline packages/eslint-plugin-ecmascript-compat/lib/features/es2020.spec.js:33]: `errors` left over after copy paste above and below here?

**Review by robatwilliams — CHANGES_REQUESTED** [review]
Looks good, thanks. Just two straightfoward things.

As a final thing, could you try it out in the example subproject, with and without options given, just for one polyfill, to make sure it works outside of unit tests? Just run `npm run lint` there. No need to commit this.

File: packages/eslint-plugin-ecmascript-compat/lib/features/es2020.spec.js
Comment (sobstel) [reply packages/eslint-plugin-ecmascript-compat/lib/features/es2020.spec.js:33]: unfortunately yes. i'll fix it.

**Comment by robatwilliams** [comment]
I looked at how eslint-plugin-compat (similar plugin for _browser_ and language features, but not syntax) allows polyfills to be specified - https://github.com/amilajack/eslint-plugin-compat#adding-polyfills

That way allows sharing between different rules - https://eslint.org/docs/latest/user-guide/configuring/configuration-files#adding-shared-settings. But like us, they only have one rule.

I don't think we need to accept polyfills from this location instead of or in addition to the rule options. It's a global location and "polyfills" is a very generic key which different plugins might interpret differently. If a user wants both plugins to share config, they can define an array in their config file and use/spread it in both locations.

**Comment by sobstel** [comment]
> As a final thing, could you try it out in the example subproject, (...)

Checked with random polyfills setup and it works ✅  
Also, I've been running it with our code base all this time, so actually double-checked.

**Comment by robatwilliams** [comment]
Great, that big codebase is the best test.

**Review by robatwilliams — APPROVED** [review]

**Comment by sobstel** [comment]
Thanks for approval. I cannot merge it on my own though. When do you think we can get it merged and released?

**Comment by robatwilliams** [comment]
Now released v2.2.0.

Thanks for your work, and please pass my thanks on to your team for continued contribution back to the plugin that you're making use of.

**Pull request merged**
